### PR TITLE
bootstrap: Start hostNetwork pods even if network plugin not ready

### PR DESCRIPTION
**Pull request by justinsb**
``` release-note
bootstrap: Start hostNetwork pods even if network plugin not ready
```

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33347)

### CODE REVIEW

**Comment by justinsb** [comment]
WIP so self-assigning for now

**Comment by justinsb** [comment]
This addresses (some of?) #32900  It is also fairly simple.

I realize this is a hard sell for 1.4.0.  But if we want to put it into 1.4 at all, maybe we can incorporate it with a flag?  Then installation tools could opt-in instead of setting `pod-cidr`.  This is of course predicated on this not being obviously wrong!

cc @bprashanth @thockin

File: cluster/gce/trusty/configure-helper.sh [hunk 1]
```diff
@@ -155,7 +155,7 @@ assemble_kubelet_flags() {
     if [ ! -z "${KUBELET_APISERVER:-}" ] && \
        [ ! -z "${KUBELET_CERT:-}" ] && \
        [ ! -z "${KUBELET_KEY:-}" ]; then
-      KUBELET_CMD_FLAGS="${KUBELET_CMD_FLAGS} --api-servers=https://${KUBELET_APISERVER} --register-schedulable=false --reconcile-cidr=false --pod-cidr=10.123.45.0/29"
+      KUBELET_CMD_FLAGS="${KUBELET_CMD_FLAGS} --api-servers=https://${KUBELET_APISERVER} --register-schedulable=false --reconcile-cidr=false"
```
Comment (freehan) [inline cluster/gce/trusty/configure-helper.sh]: If you drop `pod-cidr`, that means you have to set `reconcile-cidr`.

**Comment by freehan** [comment]
@roberthbailey Where does GKE take bootstrap parameter from? 

Bascially, we want to maintain the current setup for GKE, where pod-cidr were set on master. 
For other setups like AWS/GCE, master bootstraps without `pod-cidr` and uses node.PodCIDR once assigned.

**Comment by justinsb** [comment]
Can I get some feedback on this as fix?  If so, I propose we're early enough in the cycle that we just merge it (once it is rebased and cleaned up and passes e2e), plenty of time to revert /  fix any problems it uncovers.

**Comment by freehan** [comment]
What about the comment regarding `reconcile-cidr`?

This is a good one. Let us get it in.

**Comment by justinsb** [comment]
Thanks @freehan .  Yes, I was probably over-eager when expunging pod-cidr - I will fix it up properly :-)  Just wanted to check that we liked the general approach first!

**Comment by caseydavenport** [comment]
Haven't read the code yet, but @justinsb I thought this was already the case?  It certainly is for CNI plugins.  Is this just a kubenet fix?

**Comment by justinsb** [comment]
@caseydavenport are you sure?  We hit this pretty hard in 1.4, with the /30 -> /29 issue...  Maybe other plugins aren't correctly reporting status?  Or maybe the plugin is just ready immediately, and doesn't need to wait for a podcidr?  Check out the kubelet changes - they are pretty small...

**Comment by caseydavenport** [comment]
Yeah, my experience was with the Calico plugin, which is ready immediately and doesn't need to wait for a pod CIDR.  I _think_ I understand this PR now - it's essentially handling the case where `--configure-cbr0` is set, which will fail until a pod CIDR is configured even for host networked pods.  For cases like the Calico case where `--configure-cbr0` isn't used, we weren't hitting this before.

Sorry for the confusion!

**Comment by freehan** [comment]
In the Calico case, it is using the cni network plugin right? I believe that one already separates hostnetwork and podnetwork.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -2197,6 +2197,15 @@ func (kl *Kubelet) rejectPod(pod *api.Pod, reason, message string) {
 // can be admitted, a brief single-word reason and a message explaining why
 // the pod cannot be admitted.
 func (kl *Kubelet) canAdmitPod(pods []*api.Pod, pod *api.Pod) (bool, string, string) {
+	if rs := kl.runtimeState.networkErrors(); len(rs) != 0 {
+		hostNetwork := false
+		if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.HostNetwork {
```
Comment (dcbw) [inline pkg/kubelet/kubelet.go]: how about instead:
```
if (!podUsesHostNetwork(pod)) {
    return false, "NetworkNotReady", fmt.Sprintf(...)
}
```
We added podUsesHostNetwork() specifically to cut down on this repeated pattern :)

**Comment by dcbw** [comment]
> Yeah, my experience was with the Calico plugin, which is ready immediately and doesn't need to wait for a pod CIDR. I think I understand this PR now - it's essentially handling the case where --configure-cbr0 is set, which will fail until a pod CIDR is configured even for host networked pods. For cases like the Calico case where --configure-cbr0 isn't used, we weren't hitting this before.

@caseydavenport almost right.  Currently, if the network plugin reports an error from Status(), kubelet will block running HostNetwork=true pods that shouldn't depend on the network plugin at all.  This PR removes that erroneous inter-dependency and allows HostNetwork pods to run even when the network plugin isn't ready, since the two shouldn't have anything to do with each other.

**Comment by yujuhong** [comment]
The downside is that there is no feedback loop to prevent schedulers from assigning non-host network pods to the node. There may be a period of time after kubelet startup that it reports ready, but rejects all the (non-host network) pods. /cc @kubernetes/sig-node

**Comment by edevil** [comment]
Is there any hope of seeing this in a near future in a point release?

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (mikedanese) [inline pkg/kubelet/kubelet.go]: +1

**Review by mikedanese — COMMENTED** [review]
This change looks ok to me.

**Comment by edevil** [comment]
Hey @justinsb, can you fix the conflicts please?

**Comment by luxas** [comment]
+1 for getting this merged now

**Comment by edevil** [comment]
Well, @justinsb seems to be MIA. What would be the best course of action? Should I open a new PR with this patch rebased? Does it have other problems?

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (justinsb) [inline pkg/kubelet/kubelet.go]: Thanks - will do

**Comment by justinsb** [comment]
Sorry - email is a disaster for me!  Working on it now :-)

**Comment by justinsb** [comment]
Updated:

1) Used `podUsesHostNetwork` helper function (thanks dcbw)
2) Tweaked e2e/bringup scripts: don't pass a dummy pod-cidr, if no pod-cidr specify `--reconcile-cidr=false`, if pod-cidr specify `--reconcile-cidr=true` but only needed if kubenet.

**Comment by edevil** [comment]
@freehan Anything missing?

**Comment by yujuhong** [comment]
> Why do we want to set the flag to its default value? Just to be explicit? we don't really do that for anything else.

Sounds like we should deprecate the --reconcile-cidr flag properly if we want to set the value of the flag automatically based on --pod-cidr.

**Comment by justinsb** [comment]
I just set the flag explicitly because I thought it was clearer, given that I screwed it up first time.  I'll fix so we don't pass it unless we are changing the default.

**Comment by justinsb** [comment]
Simplified the second commit to be minimal.  Will be interesting to see if it passes all our e2es - I think the absence of `reconcile-cidr=false` in e2e_node/services/services.go is particularly surprising, but I haven't dug in.

The reason for the "dummy" CIDR (@thockin) is that during bringup etcd, apiserver, kcm and scheduler are static pods on the master.  They are hostNetwork so that they don't rely on networking, but kubelet _still_ wouldn't launch them if we were doing e.g. kubenet networking and we hadn't yet get a PodCIDR for the master.  That is from KCM, so that is a circular dependency.

The workaround was to assign a dummy CIDR 10.123.45.0/30, so that these static pods would start (and because they are hostNetwork, the dummy CIDR wouldn't matter).  Then the master would `reconcile-cidr` and get a PodCIDR.  That was fine, if a little ugly.  The problem was that kubenet wasn't updating with the PodCIDR, and was still using the 10.123.45.0/30, which meant that pods running on the master (daemonsets) would rapidly run out of IPs unless they were hostNetwork.

And so this PR side-steps the 10.123.45.0/30 trick entirely.  We just start hostNetwork pods even if we don't yet have a PodCIDR.  Once we get a PodCIDR, we can start pods that are not hostNetwork.

As before, pods that are part of bringup (i.e. everything up until KCM, or the network plugin daemonset) need to be hostNetwork.

**Comment by dcbw** [comment]
The pkg/kubelet bits LGTM

**Comment by edevil** [comment]
Is this something that will end up on the next 1.4.x or will we have to wait for 1.5?

**Comment by roberthbailey** [comment]
@edevil - if you want it in the 1.4 branch feel free to send a cherry pick pull request.

**Comment by Crassirostris** [comment]
Guys, I'm sorry for not being vocal, by this PR broke suites `gce-es-logging` and `gci-gce-es-logging` and GCL-logging test

Logging doesn't work at all after this PR!

**Comment by justinsb** [comment]
@Crassirostris I'll take a look.  Is there an open issue?

**Comment by piosz** [comment]
#35409

**Comment by justinsb** [comment]
Thanks - looking now.

**Comment by yujuhong** [comment]
FYI, this PR may have cause a lot of tests in serial suite to flake as well: #35519

**Comment by foxish** [comment]
@yujuhong @justinsb do we want to revert this PR?

**Comment by yujuhong** [comment]
I'd vote for reverting this first.

**Comment by thockin** [comment]
I would be open to a cherrypick, but it needs to be careful

**Pull request merged**